src/pke/lib/scheme/bfvrns/bfvrns-leveledshe.cpp

```diff
@@ -412,4 +726,165 @@ void LeveledSHEBFVRNS::EvalMultCoreInPlace(Ciphertext<DCRTPoly>& ciphertext, con
     ciphertext->SetDepth(ciphertext->GetDepth() + 1);
 }
 
+Ciphertext<DCRTPoly> LeveledSHEBFVRNS::EvalAutomorphism(ConstCiphertext<DCRTPoly> ciphertext, usint i,
+                                                        const std::map<usint, EvalKey<DCRTPoly>>& evalKeyMap,
+                                                        CALLER_INFO_ARGS_CPP) const {
+    const std::vector<DCRTPoly>& cv = ciphertext->GetElements();
+
+    usint N = cv[0].GetRingDimension();
+
+    std::vector<usint> vec(N);
+    PrecomputeAutoMap(N, i, &vec);
+
+    auto algo = ciphertext->GetCryptoContext()->GetScheme();
+
+    Ciphertext<DCRTPoly> result = ciphertext->Clone();
+
+    RelinearizeCore(result, evalKeyMap.at(i));
+
+    std::vector<DCRTPoly>& rcv = result->GetElements();
+
+    rcv[0] = rcv[0].AutomorphismTransform(i, vec);
+    rcv[1] = rcv[1].AutomorphismTransform(i, vec);
+
+    return result;
+}
+
+std::shared_ptr<std::vector<DCRTPoly>> LeveledSHEBFVRNS::EvalFastRotationPrecompute(
+    ConstCiphertext<DCRTPoly> ciphertext) const {
+    const auto cryptoParams = std::dynamic_pointer_cast<CryptoParametersBFVRNS>(ciphertext->GetCryptoParameters());
+    auto algo               = ciphertext->GetCryptoContext()->GetScheme();
+    std::shared_ptr<std::vector<DCRTPoly>> digits;
+
+    if (cryptoParams->GetMultiplicationTechnique() == HPSPOVERQLEVELED) {
+        digits = algo->EvalKeySwitchPrecomputeCore(ciphertext->GetElements()[1], ciphertext->GetCryptoParameters());
+    }
+    else {
```

Comment:
no else, delete "digits" and return "algo->EvalKeySwitchPrecomputeCore(c1, ciphertext->GetCryptoParameters())"